src/main/java/org/truffleruby/core/time/RubyDateFormatter.java

```diff
@@ -544,6 +568,168 @@ public static RopeBuilder formatToRopeBuilder(List<Token> compiledPattern, Zoned
+                            return formatToRopeBuilder(
+                                    compiledPattern,
+                                    dt,
+                                    zone,
+                                    context,
+                                    language,
+                                    currentNode,
+                                    errnoErrorNode).toRope();
+                        }
+
+                        appendRope = new LazyIntRope(value, UTF8Encoding.INSTANCE, 4);
+                    }
+                        break;
+
+                    case FORMAT_NANOSEC: {
+                        final int nano = dt.getNano();
+                        assert nano >= 0;
+                        assert nano < 1000000000;
+
+                        final int digits;
+
+                        if (nano < 10) {
+                            digits = 1;
+                        } else if (nano < 100) {
+                            digits = 2;
+                        } else if (nano < 1000) {
+                            digits = 3;
+                        } else if (nano < 10000) {
+                            digits = 4;
+                        } else if (nano < 100000) {
+                            digits = 5;
+                        } else if (nano < 1000000) {
+                            digits = 6;
+                        } else if (nano < 10000000) {
+                            digits = 7;
+                        } else if (nano < 100000000) {
+                            digits = 8;
+                        } else {
+                            digits = 9;
+                        }
+
+                        final LazyIntRope nanoRope = new LazyIntRope(nano, UTF8Encoding.INSTANCE, digits);
+
+                        final int padding = 6 - digits;
+
+                        if (padding == 0) {
+                            appendRope = nanoRope;
+                        } else if (padding < 0) {
+                            appendRope = new SubstringRope(UTF8Encoding.INSTANCE, nanoRope, 0, 6, 6, CodeRange.CR_7BIT);
+                        } else {
+                            appendRope = new ConcatRope(
+                                    nanoRope,
+                                    RopeConstants.paddingZeros(padding),
+                                    UTF8Encoding.INSTANCE,
+                                    CodeRange.CR_7BIT);
+                        }
+                    }
+                        break;
+
+                    default:
+                        CompilerDirectives.transferToInterpreterAndInvalidate();
+                        throw new UnsupportedOperationException();
+                }
+
+                if (rope == null) {
+                    rope = appendRope;
+                } else {
+                    rope = new ConcatRope(rope, appendRope, UTF8Encoding.INSTANCE, CodeRange.CR_UNKNOWN);
```

Comment:
> `CodeRange.CR_UNKNOWN`

That's not valid, the CodeRange of any ManagedRope must always be known, we should add an assert in the `ManagedRope` constructor for this.

I think we should use a `ConcatNode` here instead of that logic.

Reply:
Alternatively, as discussed, if we can ensure everything in the returned string is 7-bit then you can just pass CR_7BIT here (but it's a bit brittle)

Reply:
Done.